data/api.go

```diff
@@ -0,0 +1,355 @@
+package data
+
+import (
+	"crypto/rand"
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"log"
+	"net/http"
+	"net/url"
+	"time"
+
+	"github.com/google/uuid"
+
+	"github.com/gorilla/mux"
+	"github.com/tidepool-org/go-common/clients"
+	"github.com/tidepool-org/go-common/clients/mongo"
+	"github.com/tidepool-org/go-common/clients/shoreline"
+	"github.com/tidepool-org/go-common/clients/status"
+	"github.com/tidepool-org/tide-whisperer/auth"
+	"github.com/tidepool-org/tide-whisperer/store"
+)
+
+type (
+	API struct {
+		store           *store.MongoStoreClient
+		shorelineClient *shoreline.ShorelineClient
+		authClient      *auth.Client
+		perms           clients.Gatekeeper
+		schemaVersion   store.SchemaVersion
+	}
+
+	varsHandler func(http.ResponseWriter, *http.Request, map[string]string)
+
+	// so we can wrap and marshal the detailed error
+	detailedError struct {
+		Status int `json:"status"`
+		//provided to user so that we can better track down issues
+		ID              string `json:"id"`
+		Code            string `json:"code"`
+		Message         string `json:"message"`
+		InternalMessage string `json:"-"` //used only for logging so we don't want to serialize it out
+	}
+
+	//generic type as device data can be comprised of many things
+	deviceData map[string]interface{}
+)
+
+const (
+	//api logging prefix
+	DataAPIPrefix             = "api/data "
+	medtronicLoopBoundaryDate = "2017-09-01"
+	slowQueryDuration         = 0.1 // seconds
+)
+
+var (
+	errorStatusCheck       = detailedError{Status: http.StatusInternalServerError, Code: "data_status_check", Message: "checking of the status endpoint showed an error"}
+	errorNoViewPermission  = detailedError{Status: http.StatusForbidden, Code: "data_cant_view", Message: "user is not authorized to view data"}
+	errorNoPermissions     = detailedError{Status: http.StatusInternalServerError, Code: "data_perms_error", Message: "error finding permissions for user"}
+	errorRunningQuery      = detailedError{Status: http.StatusInternalServerError, Code: "data_store_error", Message: "internal server error"}
+	errorLoadingEvents     = detailedError{Status: http.StatusInternalServerError, Code: "data_marshal_error", Message: "internal server error"}
+	errorInvalidParameters = detailedError{Status: http.StatusInternalServerError, Code: "invalid_parameters", Message: "one or more parameters are invalid"}
+)
+
+func InitApi(mongoConfig mongo.Config, shoreline *shoreline.ShorelineClient, auth *auth.Client, permsClient clients.Gatekeeper, schemaV store.SchemaVersion) *API {
+	storage := store.NewMongoStoreClient(&mongoConfig)
+
```

Comment:
EnsureIndexes has disappeared ? 
```suggestion
	storage.EnsureIndexes()
```